dv/uvm/core_ibex/riscv_dv_extension/ibex_log_to_trace_csv.py

```diff
@@ -69,17 +70,17 @@ def _process_ibex_sim_log_fd(log_fd, csv_fd, full_trace=True):
             trace_entry.pc = m.group("pc")
             trace_entry.binary = m.group("bin")
             if full_trace:
+                trace_entry.operand = m.group("instr").split()[1]
+                process_trace(trace_entry)
                 # Convert the operands into ABI format for
                 # the functional coverage flow
-                operands = m.group("instr").split()[1]
-                trace_entry.operand = convert_operands_to_abi(operands)
-                process_trace(trace_entry)
+                trace_entry.operand = convert_operands_to_abi(trace_entry.operand)
```

Comment:
This code is pretty hard to follow (setting a field, calling a mutating function on the object, then setting the field again), and I think we could do it more simply. Since we don't use `process_trace` anywhere else, could we move the `convert_operands_to_abi` call inside it? Maybe something like this would be cleaner? (untested)
```python
            ...
            trace_entry.binary = m.group("bin")
            if full_trace:
                expand_trace_entry(trace_entry, m.group("instr").split()[1])

            c = RD_RE.search(line)
            ...

...

def expand_trace_entry(entry, operands):
    '''Expand a trace for a single instruction

    Add operands to the entry, converting from raw register names (x0, x1 etc.)
    to ABI names (a1, t1 etc.).

    '''
    process_imm(entry)
    entry.instr, operands = \
        convert_pseudo_instr(entry.instr, operands, entry.binary)

    # process any instructions of the form:
    # <instr> <reg> <imm>(<reg>)
    n = ADDR_RE.search(operands)
    if n:
        entry.imm = get_imm_hex_val(n.group("imm"))
        operands = ",".join([n.group("rd"), n.group("rs1"), n.group("imm")])

    # Convert the operands into ABI format for the functional coverage flow and
    # store them into the trace entry.
    entry.operand = convert_operands_to_abi(operands)

```

Reply:
Gaargh! That doesn't work because `process_imm` looks at `entry.operand` too. I probably prefer the "doubling down on not mutating stuff" approach, so `process_imm` changes to this:
```python3
...
    operands = process_imm(entry.instr, entry.pc, operands)
...

def process_imm(instr_name, pc, operands):
    """Process any immediate operand to follow RISC-V standard convention"""
    if instr_name in ['beq', 'bne', 'blt', 'bge', 'bltu', 'bgeu', 'c.beqz',
                      'c.bnez', 'beqz', 'bnez', 'bgez', 'bltz', 'blez',
                      'bgtz', 'c.j', 'j', 'c.jal', 'jal']:
        idx = operands.rfind(',')
        if idx == -1:
            imm = operands
            return str(sint_to_hex(int(imm, 16) - int(pc, 16)))

        imm = operands[idx + 1:]
        imm = str(sint_to_hex(int(imm, 16) - int(pc, 16)))
        return operands[0:idx + 1] + imm
```

The idea is you slowly move towards "compute some fields and make an (immutable) object out of them", which generally leads to much simpler code.

Reply:
hm you make a good point re: the jumble of setting fields and mutating objects - I'll restructure this along the lines of what you suggested (making sure it works ofc) and will update this PR accordingly.